[SPARK-49673][CONNECT] Increase CONNECT_GRPC_ARROW_MAX_BATCH_SIZE to 0.7 * CONNECT_GRPC_MAX_MESSAGE_SIZE

dillitz opened the pull request:
### What changes were proposed in this pull request?
Increases the default `maxBatchSize` from 4MiB * 0.7 to 128MiB (=
CONNECT_GRPC_MAX_MESSAGE_SIZE) * 0.7. This makes better use of the allowed maximum message size.
This limit is used when creating Arrow batches for the `SqlCommandResult` in the `SparkConnectPlanner` and for `ExecutePlanResponse.ArrowBatch` in `processAsArrowBatches`. This, for example, lets us return much larger `LocalRelations` in the `SqlCommandResult` (i.e., for the `SHOW PARTITIONS` command) while still staying within the GRPC message size limit.


### Why are the changes needed?
There are `SqlCommandResults` that exceed 0.7 * 4MiB.


### Does this PR introduce _any_ user-facing change?
Now support `SqlCommandResults` <= 0.7 * 128 MiB instead of only <= 0.7 * 4MiB and ExecutePlanResponses will now better use the limit of 128MiB. 


### How was this patch tested?
Existing tests.


### Was this patch authored or co-authored using generative AI tooling?
No.

dillitz committed:
increase maxBatchSize to 0.7 * grpc message size limit for sqlCommandResult

dillitz committed:
remove unused import

HyukjinKwon commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:
```diff
@@ -2498,7 +2498,7 @@ class SparkConnectPlanner(
       // Convert the results to Arrow.
       val schema = df.schema
-      val maxBatchSize = (SparkEnv.get.conf.get(CONNECT_GRPC_ARROW_MAX_BATCH_SIZE) * 0.7).toLong
+      val maxBatchSize = (ConnectCommon.CONNECT_GRPC_MAX_MESSAGE_SIZE * 0.7).toLong
```
If this is more performance in most cases, I think we should change the default value of `CONNECT_GRPC_ARROW_MAX_BATCH_SIZE` instead?

dillitz replied:
I did not want to change the value for `processAsArrowBatches`, but looking into it, I think changing the value there makes sense as well since there the ExecutePlanResponse also only contains the ArrowBatch. So increasing the limit here to 0.7 * MAX_MESSAGE_SIZE is fine as well. Let's change the config instead of only the value here then! Thanks!

dillitz replied:
Changed.

dillitz committed:
Revert "remove unused import"

This reverts commit 18a330f46362d435c7181613c9f15b268fc024fb.

dillitz committed:
Revert "increase maxBatchSize to 0.7 * grpc message size limit for sqlCommandResult"

This reverts commit a68f280253da4d8358db61a9586dbd8ee1f258b0.

dillitz committed:
change CONNECT_GRPC_ARROW_MAX_BATCH_SIZE to CONNECT_GRPC_MAX_MESSAGE_SIZE

nija-at reviewed (commented):
Any existing tests that use the new limits to ensure that everything still works?

dillitz committed:
add E2E test

dillitz commented:
> Any existing tests that use the new limits to ensure that everything still works?
I've added an E2E test that explicitly tests multiple batches and changed the static limit to 10MiB there to not create too much memory pressure.

dillitz committed:
adjust client's maxMessageSize

nija-at commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:
```diff
@@ -1566,6 +1566,25 @@ class ClientE2ETestSuite
   }
+
+  test("SPARK-49673: new batch size, multiple batches") {
+    val maxBatchSize = spark.conf.get("spark.connect.grpc.arrow.maxBatchSize").dropRight(1).toInt
```
This drops the limit by one order of magnitude. Could we run this at 80% of the limit?

dillitz replied:
Sorry not sure what you mean here. I've decreased both the GRPC_MAX_MESSAGE_SIZE in the client and the maxBatchSize on the server to 10MiB to simulate the batches being close to the message size limit without having to process `range(128*1024*1024)` on the cluster. Are you asking me to replace this with spark.connect.grpc.arrow.maxBatchSize = spark.connect.grpc.arrow.maxBatchSize = CONNECT_GRPC_MAX_MESSAGE_SIZE = 0.8 * 128MiB?

dillitz replied:
Ah, I think there might be a misunderstanding here. The dropRight is used to drop the "b" character from the "1269760b" String returned by the spark.conf.get("spark.connect.grpc.arrow.maxBatchSize") call 😄

nija-at replied:
Ah, never mind then :)

dillitz committed:
linter

HyukjinKwon commented:
Merged to master.
